modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -606,14 +608,29 @@ def downsample_batch(
         assert self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE
 
         self._downsampler.init_downsampler()
+
+        if self._downsampler.requires_coreset_supporting_module:
+            # enable the embedding recorder to keep track of last layer embedding. The embeddings are stored
+            # in self._model.model.embedding_recorder.embedding
+            assert isinstance(self._model.model, CoresetSupportingModule)
+            self._model.model.embedding_recorder.start_recording()
+
         big_batch_output = self._model.model(data)
-        self._downsampler.inform_samples(sample_ids, big_batch_output, target)
+
+        # supply the embeddings if required by the downsampler
+        if self._downsampler.requires_coreset_supporting_module:
+            embeddings = self._model.model.embedding_recorder.embedding
+            self._model.model.embedding_recorder.end_recording()
+        else:
+            embeddings = None
+
+        self._downsampler.inform_samples(sample_ids, big_batch_output, target, embeddings)
         # TODO(#218) Persist information on the sample IDs/weights when downsampling is performed
         selected_indexes, weights = self._downsampler.select_points()
         selected_data, selected_target = get_tensors_subset(selected_indexes, data, target, sample_ids)
         sample_ids, data, target = selected_indexes, selected_data, selected_target
         # TODO(#219) Investigate if we can avoid 2 forward passes
-        return data, sample_ids, target, weights
+        return data, sample_ids, target, weights.to(self._device)
```

Comment:
Line 390 is for STB, where we retrieve the samples from the disk. This function is for BTS, where weights are in memory. Hence, we need both `.to` since they operate on different tensors.